pkg/controller/installation/windows/calico_windows_upgrader.go

```diff
@@ -0,0 +1,414 @@
+)
+
+var (
+	windowsLog = logf.Log.WithName("windows_upgrader")
+
+	defaultMaxUnavailable = 1
+)
+
+type CalicoWindowsUpgrader interface {
+	UpdateConfig(install *operatorv1.InstallationSpec)
+	Start(ctx context.Context)
+}
+
+// calicoWindowsUpgrader helps manage the upgrade of Calico Windows nodes.
+// It works in conjunction with the CalicoUpgrade service running on each node.
+type calicoWindowsUpgrader struct {
+	clientset     kubernetes.Interface
+	client        client.Client
+	statusManager status.StatusManager
+	nodeIndexer   cache.SharedIndexInformer
+	syncPeriod    time.Duration
+	installChan   chan *operatorv1.InstallationSpec
+	install       *operatorv1.InstallationSpec
+}
+
+type calicoWindowsUpgraderOption func(*calicoWindowsUpgrader)
+
+// calicoWindowsUpgraderSyncPeriod sets the sync period for the
+// calicoWindowsUpgrader.
+func CalicoWindowsUpgraderSyncPeriod(syncPeriod time.Duration) calicoWindowsUpgraderOption {
+	return func(w *calicoWindowsUpgrader) {
+		w.syncPeriod = syncPeriod
+	}
+}
+
+// NewCalicoWindowsUpgrader creates a Calico Windows upgrader.
+func NewCalicoWindowsUpgrader(cs kubernetes.Interface, c client.Client, indexer cache.SharedIndexInformer, statusManager status.StatusManager, options ...calicoWindowsUpgraderOption) CalicoWindowsUpgrader {
+	w := &calicoWindowsUpgrader{
+		clientset:     cs,
+		client:        c,
+		statusManager: statusManager,
+		nodeIndexer:   indexer,
+		syncPeriod:    10 * time.Second,
+		installChan:   make(chan *operatorv1.InstallationSpec, 100),
+	}
+
+	for _, o := range options {
+		o(w)
+	}
+	return w
+}
+
+// UpdateConfig updates the calicoWindowsUpgrader's installation config.
+func (w *calicoWindowsUpgrader) UpdateConfig(install *operatorv1.InstallationSpec) {
+	w.installChan <- install
+}
+
+// getNodeUpgradeStatus checks the nodes from its indexer and determines whether
+// the nodes are:
+// - pending: Node does not have the expected variant and/or version. No upgrade label.
+// - inProgress: Node does not have the expected variant and/or version. It has the upgrade in-progress label.
+// - inSync: Node has the expected variant and version. It does not have the upgrade in-progress label.
+// This returns an error (if any) and maps of the nodes that are pending,
+// inProgress, or inSync.
+func (w *calicoWindowsUpgrader) getNodeUpgradeStatus() (map[string]*corev1.Node, map[string]*corev1.Node, map[string]*corev1.Node, error) {
+	pending := make(map[string]*corev1.Node)
+	inSync := make(map[string]*corev1.Node)
+	inProgress := make(map[string]*corev1.Node)
+	expectedVersion := w.getExpectedVersion()
+
+	for _, obj := range w.nodeIndexer.GetIndexer().List() {
+		node, ok := obj.(*corev1.Node)
+		if !ok {
+			return nil, nil, nil, fmt.Errorf("Never expected index to have anything other than a Node object: %v", obj)
+		}
+
+		if node.Labels[corev1.LabelOSStable] != "windows" {
+			continue
+		}
+
+		windowsLog.V(1).Info(fmt.Sprintf("Processing node %v", node.Name))
```

Comment:
This log should probably be debug, otherwise it will get quite spammy.

Reply:
I'll check again but I thought the default zap log level was info and so V(1) logs will be debug.

Reply:
Oh yeah I forgot we're using zap in the operator.